sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ProductAggSuite.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.AttributeReference
+import org.apache.spark.sql.types.DoubleType
+
+
+class ProductAggSuite extends SparkFunSuite {
+  val input = AttributeReference("product", DoubleType, nullable = true)()
+  val evaluator = DeclarativeAggregateEvaluator(Product(input), Seq(input))
+
+  test("empty buffer") {
+    assert(evaluator.initialize() === InternalRow(null))
+  }
+
+  test("update") {
+    val result = evaluator.update(
+      InternalRow(-2.0),
+      InternalRow(3.0),
+      InternalRow(-5.0),
+      InternalRow(7.0))
+    assert(result === InternalRow(210.0))
+  }
+
+  test("update - with nulls") {
+    val result1 = evaluator.update(
+      InternalRow(null),
+      InternalRow(11.0),
+      InternalRow(null),
+      InternalRow(13.0))
+    assert(result1 === InternalRow(143.0))
+
+    val result2 = evaluator.update(
+      InternalRow(null),
+      InternalRow(null))
+    assert(result2 === InternalRow(null))
+  }
+
+  test("merge") {
+    // Empty
+    val p0 = evaluator.initialize()
+    assert(evaluator.merge(p0) === InternalRow(null))
+
+    // Singleton
+    val p1 = evaluator.update(InternalRow(7.0), InternalRow(11.0))
+    assert(evaluator.merge(p1) === p1)
+
+    // Pair
+    val p2 = evaluator.update(InternalRow(17.0), InternalRow(19.0))
+    assert(evaluator.merge(p1, p2) === InternalRow((7 * 11 * 17 * 19).toDouble))
```

Comment:
nit: `7.0 * 11 * 17 * 19`

Reply:
I prefer the original because it's a *slightly* stronger test of round-off errors. The original computes the product using integer arithmetic *before* conversion to `Double`, and confirms that this is consistent with `product` which uses conversion of each element individually to a `Double`.

Reply:
(I agree with that)